### ISSUE TITLE
Verifying data integrity when copying blobs

### ISSUE THREAD

**Issue by chrisfrank** [googleapis/python-storage#422]
**Is your feature request related to a problem? Please describe.**

In our data processing pipelines, we see infrequent `google.resumable_media.common.DataCorruption` errors when calling `Blob.download_as_bytes` -- which is helpful, since it lets us retry until our data arrives intact.

These pipelines also use `Blob.rewrite` to copy large files from one GCS bucket to another, and I'm a little concerned that we've never seen a single `DataCorruption` error from `Blob.rewrite`. Have we seen no errors because this method works 100% of the time, or because it does not verify checksums?

 **Describe the solution you'd like**

We'd like to be able to ensure that requests to copy a blob will fail when the checksums at the source and destination do not match.

 **Describe alternatives you've considered**

- An earlier version of this pipeline used `Bucket.copy_blob` instead of `Blob.rewrite`, but we switched to `rewrite` after seeing log warnings about `copy_blob` being inefficient for large files. (I'm not sure if `copy_blob` verifies checksums either.)
- Our pipeline could download the blob locally with checksum verification, then upload it to the destination bucket with checksum verification, but this would take longer and require nontrivial transient storage space.

**Comment by frankyn** [comment]
Hi @chrisfrank 

Thanks for your question. 

1. Copy and Rewrite operations occur in the GCS backend and the service will validate checksums as it copies data for correctness.
2. `google.resumable_media.common.DataCorruption` occurs only when a mismatch occurs on download due to transfer corruption, e.g. solar rays, random events, network issues before data is received by the client.

Hope this answers your question.

**Comment by chrisfrank** [comment]
Hi @frankyn,

Thanks very much for the fast response! This is exactly what I was hoping to hear, but I actually opened this feature request in response to GCP Support informing my team that GCS does _not_ validate checksums in this case. I've noted the inconsistency in our support ticket, and if it's okay, I'd like to leave this feature request open until we have a consistent account of the system's behavior.

**Comment by frankyn** [comment]
Thanks for calling that out. My understanding may be incorrect but will check-in.

Could you forward me the GCS Support id over email to coderfrank@google.com?

Apologies for any confusion here! If you're okay with closing out this issue, I'll work internally to follow-up on your support ticket instead.

**Comment by chrisfrank** [comment]
Sure thing, sent!

**Comment by frankyn** [comment]
Hi @chrisfrank, (we will provide this info over the support channel as well and sharing it for general audience while we update our public documentation to make it clear) 

We reached out to the backend team and they confirmed that for server side operations such as rewrite and copy, objects are validated whenever they're moved to prevent corruption. 

You do need to validate checksums on uploads and downloads to and from GCS only (as it's implemented now).

 Thanks for the questions. Closing this issue.
